modules/supervisord/collect.go

```diff
@@ -0,0 +1,172 @@
+	}
+	s.cleanupCache()
+}
+
+func (s *Supervisord) resetCache() {
+	for _, procs := range s.cache {
+		for name := range procs {
+			procs[name] = false
+		}
+	}
+}
+
+func (s *Supervisord) cleanupCache() {
+	for group, procs := range s.cache {
+		for name, ok := range procs {
+			if !ok {
+				s.removeProcessFromCharts(group, name)
+				delete(s.cache[group], name)
+			}
+		}
+		if len(s.cache[group]) == 0 {
+			s.removeProcessGroupCharts(group)
+			delete(s.cache, group)
+		}
+	}
+}
+
+func calcProcessUptime(p processStatus) int64 {
+	if !isProcRunning(p) {
+		return 0
+	}
+	return int64(p.now - p.start)
+}
+
+func calcProcessDowntime(p processStatus) int64 {
+	if isProcRunning(p) || p.stop == 0 {
+		return 0
+	}
+	return int64(p.now - p.stop)
+}
+
+func (s *Supervisord) addProcessGroupCharts(p processStatus) {
+	charts := newProcGroupCharts(p.group)
+	if err := s.Charts().Add(*charts...); err != nil {
+		s.Warning(err)
+	}
+}
+
+func (s *Supervisord) addProcessToCharts(p processStatus) {
+	id := procID(p)
+	for _, c := range *s.Charts() {
+		var dimID string
+		switch c.ID {
+		case fmt.Sprintf(groupProcessesStateCodeChartTmpl.ID, p.group):
+			dimID = id + "_state_code"
+		case fmt.Sprintf(groupProcessesExitStatusChartTmpl.ID, p.group):
+			dimID = id + "_exit_status"
+		case fmt.Sprintf(groupProcessesUptimeChartTmpl.ID, p.group):
+			dimID = id + "_uptime"
+		case fmt.Sprintf(groupProcessesDowntimeChartTmpl.ID, p.group):
+			dimID = id + "_downtime"
+		default:
+			continue
+		}
+		dim := &module.Dim{ID: dimID, Name: p.name}
+		if err := c.AddDim(dim); err != nil {
+			s.Warning(err)
+			return
+		}
+		c.MarkNotCreated()
+	}
+}
+
+func (s *Supervisord) removeProcessGroupCharts(group string) {
+	prefix := "group_" + group
+	for _, c := range *s.Charts() {
+		if strings.HasPrefix(c.ID, prefix) {
+			c.MarkRemove()
+			c.MarkNotCreated()
+		}
+	}
+}
+
+func (s *Supervisord) removeProcessFromCharts(group, name string) {
+	id := procID(processStatus{name: name, group: group})
+	for _, c := range *s.Charts() {
+		var dimID string
+		switch c.ID {
+		case fmt.Sprintf(groupProcessesStateCodeChartTmpl.ID, group):
+			dimID = id + "_state_code"
+		case fmt.Sprintf(groupProcessesExitStatusChartTmpl.ID, group):
+			dimID = id + "_exit_status"
+		case fmt.Sprintf(groupProcessesUptimeChartTmpl.ID, group):
+			dimID = id + "_uptime"
+		case fmt.Sprintf(groupProcessesDowntimeChartTmpl.ID, group):
+			dimID = id + "_downtime"
+		default:
+			continue
+		}
+		if err := c.MarkDimRemove(dimID, true); err != nil {
+			s.Warning(err)
+			return
+		}
+		c.MarkNotCreated()
+	}
+}
+
+func procID(p processStatus) string {
+	return fmt.Sprintf("group_%s_process_%s", p.group, p.name)
+}
+
+func isProcRunning(p processStatus) bool {
+	// http://supervisord.org/subprocess.html#process-states
+	// STOPPED  (0)
+	// STARTING (10)
+	// RUNNING (20)
+	// BACKOFF (30)
+	// STOPPING (40)
+	// EXITED (100)
+	// FATAL (200)
+	// UNKNOWN (1000)
+	return p.state == 20
```

Comment:
Seems this is not correct, should be `STARTING || RUNNING || STOPPING` 🤔

Reply:
@ekexcello what do you think? Uptime [`STARTING`, `STOPPING`], downtime [`STOPPING`, `STARTING`]

Reply:
Hello @ilyam8 
Sorry, but looking at function code I can't get the question, could you please rephrase it?

Reply:
Do we consider process as running if its state is `STARTING`/`STOPPING`?
We need this to have begin/end points to calc uptime/downtime.

Reply:
I think yes. And i need to update the code 😄

Reply:
On the other hangs, if the process is hanging in `STARTING`.. We don't want to treat it as running and show its uptime, it is still downtime 🤔

Reply:
I agree that process in "starting" state is not yet running, and "stopping" is for sure not up. However, I'll also ask opinion of my colleagues, will update you later providing their opinion, too

Reply:
Keep in mind that visualising `state code` and `exit status` until we don't do averaging - only for real-time data. Averaging is ok for some metrics (number of running processes) and doesn't work at all for other (e.g. `state code`)

That is the data for the last 30 minutes (actual state codes values: 20, 100).

<img width="1346" alt="Screenshot 2021-02-16 at 16 25 58" src="https://user-images.githubusercontent.com/22274335/108068836-a95e1f80-7073-11eb-8ebe-045f4eb6a1ae.png">

Reply:
Charts look promising! However, it would be good to add explanation with state code values if possible.

> I agree that process in "starting" state is not yet running, and "stopping" is for sure not up. 
Opinion of my colleagues is very same: "starting means not running".

Reply:
> However, it would be good to add explanation with state code values if possible.

Yeah, i will add it. I need to update `dashboard_info.js` file in the netdata repo...

---

> stopping means running - it's not yet stopped

ok, so lets open new issue (i will do it) and discuss all fixes/improvements in there. My goal was to provide minimal requested functionality and improve later when we have it refined.